addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeGatewayHandler.java

```diff
@@ -0,0 +1,575 @@
+import java.util.Collection;
+import java.util.Date;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.library.types.DateTimeType;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingRegistry;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.thing.binding.builder.ThingStatusInfoBuilder;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.mihome.MiHomeBindingConstants;
+import org.openhab.binding.mihome.internal.api.constants.DeviceConstants;
+import org.openhab.binding.mihome.internal.api.constants.DeviceTypesConstants;
+import org.openhab.binding.mihome.internal.api.constants.JSONResponseConstants;
+import org.openhab.binding.mihome.internal.api.manager.FailingRequestHandler;
+import org.openhab.binding.mihome.internal.api.manager.FailingRequestHandlerImpl;
+import org.openhab.binding.mihome.internal.api.manager.MiHomeApiConfiguration;
+import org.openhab.binding.mihome.internal.api.manager.MiHomeApiManager;
+import org.openhab.binding.mihome.internal.api.manager.MiHomeApiManagerImpl;
+import org.openhab.binding.mihome.internal.api.manager.ThingCallback;
+import org.openhab.binding.mihome.internal.rest.RestClient;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * Handler for the Mi|Home Gateway (MIHO001).
+ * The gateway is a central communication hub between the individual devices.
+ *
+ * @author Mihaela Memova - Initial contribution
+ * @author Svilen Valkanov - Added methods to get information about subdevices
+ */
+public class MiHomeGatewayHandler extends BaseBridgeHandler {
+
+    /** email address used for the user's registration */
+    private String username;
+
+    /** user's password */
+    private String password;
+
+    /** Gateway code (10 capital letters) written at the bottom of the gateway device */
+    private String gatewayCode;
+
+    /** ID of the gateway device (received after registration) */
+    private int gatewayId;
+
+    /** Label of the gateway device */
+    private String gatewayLabel;
+
+    /** Pattern used to check if a given String represents an email address */
+    public static final String EMAIL_PATTERN = "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@"
```

Comment:
Yes thats great!

Just enough to make sure it's most likely an e-mail address, it's not in the users interest to do it wrong, it's just to aid the user so that they should enter an e-mail address.